grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerQueryImpl.java

```diff
@@ -376,15 +396,26 @@ public void addAtomConstraints(Set<? extends Atomic> cstrs){
         if (atoms.size() == 1) return atoms;
 
         //pass relations or rule-resolvable types and resources
-        Set<Atom> selectedAtoms = atoms.stream()
+        Set<Atom> atomsToSelect = atoms.stream()
                 .filter(atom -> (atom.isSelectable() || atom.isRuleResolvable()))
                 .collect(Collectors.toSet());
 
-        //order by variables
         Set<Atom> orderedSelection = new LinkedHashSet<>();
-        getVarNames().forEach(var -> orderedSelection.addAll(selectedAtoms.stream()
-                .filter(atom -> atom.containsVar(var))
-                .collect(Collectors.toSet())));
+
+        Atom atom = findFirstJoinable(atomsToSelect);
+        orderedSelection.add(atom);
+        atomsToSelect.remove(atom);
```

Comment:
You can remove these two lines (they are repeated in the loop), if you move the line:
```
atom = findNextJoinable(atomsToSelect, joinedVars);
```
to after the conditional.